File: ios/PowerPlayHockeyCore/Sources/PowerPlayHockeyCore/Model/Prefs.swift
import Foundation

/// Tiny wrapper over UserDefaults for the audio settings (0...100, 0 = off).
/// UserDefaults.standard is process-global, so unlike the Android version
/// this needs no Context argument.
enum Prefs {
    private static let keySfxVolume = "sfx_volume"
    private static let keyMusicVolume = "music_volume"

    static let defaultSfx = 100
    static let defaultMusic = 70

    static func sfxVolume() -> Int {
        clamp(UserDefaults.standard.object(forKey: keySfxVolume) as? Int ?? defaultSfx)
    }

    static func musicVolume() -> Int {
        clamp(UserDefaults.standard.object(forKey: keyMusicVolume) as? Int ?? defaultMusic)
    }

    static func setSfxVolume(_ v: Int) {
        UserDefaults.standard.set(clamp(v), forKey: keySfxVolume)
    }

    static func setMusicVolume(_ v: Int) {
        UserDefaults.standard.set(clamp(v), forKey: keyMusicVolume)
    }

    static func soundEnabled() -> Bool { sfxVolume() > 0 }
    private static let keyArena = "arena_type"
    private static let keyTournament = "tournament_state"

    static func arenaType() -> ArenaType {
        guard let raw = UserDefaults.standard.string(forKey: keyArena),
              let type = ArenaType(rawValue: raw) else {
            return .indoor
        }
        return type
    }

    static func setArenaType(_ type: ArenaType) {
        UserDefaults.standard.set(type.rawValue, forKey: keyArena)
    }

    static func saveTournament(_ state: TournamentState) {
        if let data = try? JSONEncoder().encode(state) {
            UserDefaults.standard.set(data, forKey: keyTournament)
        }
    }

    static func getTournament() -> TournamentState? {
        guard let data = UserDefaults.standard.data(forKey: keyTournament),
              let state = try? JSONDecoder().decode(TournamentState.self, from: data) else {
            return nil
        }
        return state
    }

    static func clearTournament() {
        UserDefaults.standard.removeObject(forKey: keyTournament)
    }

    private static func clamp(_ v: Int) -> Int { min(max(v, 0), 100) }
}

File: app/src/main/java/com/tablehockey/game/model/Prefs.kt
package com.tablehockey.game.model

import android.content.Context

/** Tiny wrapper over SharedPreferences for the audio settings (0..100, 0 = off). */
object Prefs {
    private const val NAME = "settings"
    private const val KEY_SFX_VOLUME = "sfx_volume"
    private const val KEY_MUSIC_VOLUME = "music_volume"

    const val DEFAULT_SFX = 100
    const val DEFAULT_MUSIC = 70

    private fun prefs(ctx: Context) = ctx.getSharedPreferences(NAME, Context.MODE_PRIVATE)

    fun sfxVolume(ctx: Context): Int = prefs(ctx).getInt(KEY_SFX_VOLUME, DEFAULT_SFX).coerceIn(0, 100)
    fun musicVolume(ctx: Context): Int = prefs(ctx).getInt(KEY_MUSIC_VOLUME, DEFAULT_MUSIC).coerceIn(0, 100)

    fun setSfxVolume(ctx: Context, v: Int) {
        prefs(ctx).edit().putInt(KEY_SFX_VOLUME, v.coerceIn(0, 100)).apply()
    }

    fun setMusicVolume(ctx: Context, v: Int) {
        prefs(ctx).edit().putInt(KEY_MUSIC_VOLUME, v.coerceIn(0, 100)).apply()
    }

    fun soundEnabled(ctx: Context): Boolean = sfxVolume(ctx) > 0
    fun musicEnabled(ctx: Context): Boolean = musicVolume(ctx) > 0

    private const val KEY_TOURNAMENT = "tournament_state"
    private const val KEY_ARENA = "arena_type"

    fun arenaType(ctx: Context): ArenaType {
        val name = prefs(ctx).getString(KEY_ARENA, ArenaType.INDOOR.name)
        return try { ArenaType.valueOf(name ?: ArenaType.INDOOR.name) } catch (e: Exception) { ArenaType.INDOOR }
    }

    fun setArenaType(ctx: Context, type: ArenaType) {
        prefs(ctx).edit().putString(KEY_ARENA, type.name).apply()
    }

    fun saveTournament(ctx: Context, state: TournamentState) {
        prefs(ctx).edit().putString(KEY_TOURNAMENT, state.toJson()).apply()
    }

    fun getTournament(ctx: Context): TournamentState? {
        val str = prefs(ctx).getString(KEY_TOURNAMENT, null) ?: return null
        return TournamentState.fromJson(str)
    }

    fun clearTournament(ctx: Context) {
        prefs(ctx).edit().remove(KEY_TOURNAMENT).apply()
    }
}
